### PR TITLE
Move counters collection out to a separate node process

### PR DESCRIPTION
**Pull request by ranasaria**
This change moves execution of performance counters collection in a separate local 'node' process.
This serves two purposes:
 * 1. Our measurement is not tainted by the cpu/memory usage of collection code itself.
 * 2. The host code being measured need not be 'node' target. It could be an electron target for example.

Also added @types for a lot of dependencies and made some minor code cleanup fixes.

### CODE REVIEW

File: src/counters.ts [hunk 1]
```diff
@@ -163,6 +168,15 @@ export class Counters {
 	@IsBoolean()
 	public readonly includeParent: boolean;
 
+	/**
+	 * whether we should skip process process from the list of processes that we are tracking.
```
Comment (Charles-Gagnon) [inline src/counters.ts]: >process process [](start = 27, length = 15)

double process

Comment (ranasaria) [reply src/counters.ts]: good catch. Fixed.

File: src/collectionProcess.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+*/
+'use strict';
+import debugLogger = require('debug');
+
+import { CollectorMethodType, CollectorResponse, ConstructCountersType, isConstructCountersType } from './remoteCollector';
+
+import { Counters } from './counters';
+import { jsonDump } from './utils';
+
+const trace = debugLogger('adstest:remotecollector:trace');
```
Comment (Charles-Gagnon) [inline src/collectionProcess.ts]: >remotecollector [](start = 35, length = 15)

collectionProcess?

Comment (ranasaria) [reply src/collectionProcess.ts]: good catch on the copy/paste bug.

File: src/collectionProcess.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+console.log("Process: " + process.argv[1] + " has now started.");
+let counters: Counters;
```
Comment (Charles-Gagnon) [inline src/collectionProcess.ts:19]: You use console.log in some places but trace/debug in others? Why not stay consistent?

File: src/collectionProcess.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+const processMessage: NodeJS.MessageListener = async (message: any) => {
+	trace('message from parent:', jsonDump(message));
+	//console.log(`is message ConstructCountersType: ${message is ConstructCountersType}`)
+	if (isConstructCountersType(message)) {
```
Comment (Charles-Gagnon) [inline src/collectionProcess.ts:22]: remove

File: src/collectionProcess.ts (hunk 2 above)
Comment (ranasaria) [reply src/collectionProcess.ts]: Fixed.

File: src/collectionProcess.ts (hunk 4 above)
Comment (ranasaria) [reply src/collectionProcess.ts:22]: removed, thanks,

File: src/remoteCollector.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+	}
+
+	public async stop(): Promise<void> {
+		await this.startPromise; //make sure that start operation had finished.
+		this.stopPromise = new DeferredPromise<void>();
+		this.collectorProcess.send(CollectorMethodType.StopCounters); // send the stop counters message to the collector child process
+		trace(`message sent to collector process: ${jsonDump(CollectorMethodType.StopCounters)}`);
+		await this.stopPromise; // await the stop of counters collection to finish.
+		this.collectorProcess.kill(); // kill the child collection process.
```
Comment (Charles-Gagnon) [inline src/remoteCollector.ts:109]: For these where we're waiting on messages to come back consider having a timeout so that if the child process gets stuck or something else happens we don't sit there spinning forever

File: src/charts.ts [hunk 6]
```diff
@@ -37,6 +42,27 @@ export interface LineData {
 export async function writeChartToFile(xData: number[], lines: LineData[], fileType: string = 'png', startTimeStamp: number = 0, xAxisLabel: string = 'elapsed(ms)', file: string = null, title: string = null): Promise<Buffer> {
 	xAxisLabel = `${xAxisLabel}, Start time:${toDateTimeString(startTimeStamp)}`;
 	xData = xData.map(x => x - xData[0]);
+	const yAxes: ChartYAxe = {
```
Comment (Charles-Gagnon) [inline src/charts.ts:45]: Not sure what these changes have to do with moving collection out to separate process. It's usually preferred to try and keep unrelated changes in separate PRs to help reduce review randomization.

**Review by Charles-Gagnon — APPROVED** [review]
:shipit:

File: src/charts.ts (hunk 6 above)
Comment (ranasaria) [reply src/charts.ts:45]: these have to do with pulling in @types for dependencies. I agree nevertheless on keeping these kinds of changes separate, so will bear that in mind in future.

File: src/remoteCollector.ts (hunk 5 above)
Comment (ranasaria) [reply src/remoteCollector.ts:109]: The child process is tied to the main process, so if and when the main process exits, the child process will exit. During the stop operation, child process is massaging all the data collected and producing charts for them which over a long stress run could be sizable processing. So it is expected that stop operation can take a long time. 

A timeout for counters construction and start operation might be a good idea where if those operations take longer than a timeout amount of time, then we may want to bail the entire test run, but currently I think I would rather not bail the stress run but just log the fact that counters collection failed. I need to think a little more on this scenario before putting the correct defensive code here. Getting some runs under the belt with this functionality will make the right call here.

Comment (ranasaria) [reply src/remoteCollector.ts:109]: So for now will get this code in its current form.

**Review by anthonydresser — APPROVED** [review]
